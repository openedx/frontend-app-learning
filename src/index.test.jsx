import {
  APP_INIT_ERROR, APP_READY, subscribe,
} from '@edx/frontend-platform';

// Jest needs this for module resolution
import * as app from '.'; // eslint-disable-line @typescript-eslint/no-unused-vars

// These need to be var not let so they get hoisted
// and can be used by jest.mock (which is also hoisted)
var mockRender; // eslint-disable-line no-var
var mockCreateRoot; // eslint-disable-line no-var
jest.mock('react-dom/client', () => {
  mockRender = jest.fn();
  mockCreateRoot = jest.fn(() => ({
    render: mockRender,
  }));

  return ({
    createRoot: mockCreateRoot,
  });
});

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  StrictMode: 'React Strict Mode',
}));

jest.mock('@edx/frontend-platform', () => ({
  APP_READY: 'app-is-ready-key',
  APP_INIT_ERROR: 'app-init-error',
  subscribe: jest.fn(),
  initialize: jest.fn(),
  mergeConfig: jest.fn(),
  getConfig: () => ({
    FAVICON_URL: 'favicon-url',
  }),
  ensureConfig: jest.fn(),
}));

jest.mock('./generic/PageNotFound', () => 'Page Not Found');
jest.mock('./course-home/goal-unsubscribe', () => 'Goal Unsubscribe');
jest.mock('./courseware/CoursewareRedirectLandingPage', () => 'Courseware Redirect Landing Page');
jest.mock('./preferences-unsubscribe', () => 'Preferences Unsubscribe');
jest.mock('./generic/CourseAccessErrorPage', () => 'Course Access Error Page');
jest.mock('./tab-page', () => ({ TabContainer: 'Tab Container' }));
jest.mock('./course-home/outline-tab', () => 'Outline Tab');
jest.mock('./course-home/live-tab/LiveTab', () => 'Live Tab');
jest.mock('./course-home/dates-tab', () => 'Dates Tab');
jest.mock('./course-home/discussion-tab/DiscussionTab', () => 'Discussion Tab');
jest.mock('./course-home/progress-tab/ProgressTab', () => 'Progress Tab');
jest.mock('./courseware/course/course-exit', () => ({ CourseExit: 'Course Exit' }));
jest.mock('./courseware', () => 'Courseware Container');

describe('app registry', () => {
  let getElement;

  beforeEach(() => {
    mockCreateRoot.mockClear();
    mockRender.mockClear();

    getElement = window.document.getElementById;
    window.document.getElementById = jest.fn(id => ({ id }));
  });
  afterAll(() => {
    window.document.getElementById = getElement;
  });

  test('subscribe: APP_READY.  links App to root element', () => {
    const callArgs = subscribe.mock.calls[0];
    expect(callArgs[0]).toEqual(APP_READY);
    callArgs[1]();
  });
  test('subscribe: APP_INIT_ERROR.', () => {
    const callArgs = subscribe.mock.calls[1];
    expect(callArgs[0]).toEqual(APP_INIT_ERROR);
    const error = { message: 'test-error-message' };
    callArgs[1](error);
  });
});
