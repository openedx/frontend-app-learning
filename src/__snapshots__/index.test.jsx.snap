// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app registry subscribe: APP_INIT_ERROR.  snapshot: displays an ErrorPage to root element 1`] = `
<React Strict Mode>
  <ErrorPage
    message="test-error-message"
  />
</React Strict Mode>
`;

exports[`app registry subscribe: APP_READY.  links App to root element 1`] = `
<React Strict Mode>
  <AppProvider
    store={
      {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(Symbol.observable): [Function],
      }
    }
  >
    <HelmetWrapper
      defer={true}
      encodeSpecialCharacters={true}
    >
      <link
        href="favicon-url"
        rel="shortcut icon"
        type="image/x-icon"
      />
    </HelmetWrapper>
    <PathFixesProvider>
      <NoticesProvider>
        <UserMessagesProvider>
          <div
            className="app-container"
          >
            <Routes>
              <Route
                element={
                  <PageWrap>
                    <Page Not Found />
                  </PageWrap>
                }
                path="*"
              />
              <Route
                element={
                  <PageWrap>
                    <Goal Unsubscribe />
                  </PageWrap>
                }
                path="/goal-unsubscribe/:token"
              />
              <Route
                element={
                  <PageWrap>
                    <Courseware Redirect Landing Page />
                  </PageWrap>
                }
                path="/redirect/*"
              />
              <Route
                element={
                  <PageWrap>
                    <Preferences Unsubscribe />
                  </PageWrap>
                }
                path="/preferences-unsubscribe/:userToken/:updatePatch"
              />
              <Route
                element={
                  <DecodePageRoute>
                    <Course Access Error Page />
                  </DecodePageRoute>
                }
                path="/course/:courseId/access-denied"
              />
              <Route
                element={
                  <DecodePageRoute>
                    <Tab Container
                      fetch={[Function]}
                      slice="courseHome"
                      tab="outline"
                    >
                      <Outline Tab />
                    </Tab Container>
                  </DecodePageRoute>
                }
                path="/course/:courseId/home"
              />
              <Route
                element={
                  <DecodePageRoute>
                    <Tab Container
                      fetch={[Function]}
                      slice="courseHome"
                      tab="lti_live"
                    >
                      <Live Tab />
                    </Tab Container>
                  </DecodePageRoute>
                }
                path="/course/:courseId/live"
              />
              <Route
                element={
                  <DecodePageRoute>
                    <Tab Container
                      fetch={[Function]}
                      slice="courseHome"
                      tab="dates"
                    >
                      <Dates Tab />
                    </Tab Container>
                  </DecodePageRoute>
                }
                path="/course/:courseId/dates"
              />
              <Route
                element={
                  <DecodePageRoute>
                    <Tab Container
                      fetch={[Function]}
                      slice="courseHome"
                      tab="discussion"
                    >
                      <Discussion Tab />
                    </Tab Container>
                  </DecodePageRoute>
                }
                path="/course/:courseId/discussion/:path/*"
              />
              <Route
                element={
                  <DecodePageRoute>
                    <Tab Container
                      fetch={[Function]}
                      isProgressTab={true}
                      slice="courseHome"
                      tab="progress"
                    >
                      <Progress Tab />
                    </Tab Container>
                  </DecodePageRoute>
                }
                path="/course/:courseId/progress/:targetUserId/"
              />
              <Route
                element={
                  <DecodePageRoute>
                    <Tab Container
                      fetch={[Function]}
                      isProgressTab={true}
                      slice="courseHome"
                      tab="progress"
                    >
                      <Progress Tab />
                    </Tab Container>
                  </DecodePageRoute>
                }
                path="/course/:courseId/progress"
              />
              <Route
                element={
                  <DecodePageRoute>
                    <Tab Container
                      fetch={[Function]}
                      slice="courseware"
                      tab="courseware"
                    >
                      <Course Exit />
                    </Tab Container>
                  </DecodePageRoute>
                }
                path="/course/:courseId/course-end"
              />
              <Route
                element={
                  <DecodePageRoute>
                    <Courseware Container />
                  </DecodePageRoute>
                }
                path="/course/:courseId/:sequenceId/:unitId"
              />
              <Route
                element={
                  <DecodePageRoute>
                    <Courseware Container />
                  </DecodePageRoute>
                }
                path="/course/:courseId/:sequenceId"
              />
              <Route
                element={
                  <DecodePageRoute>
                    <Courseware Container />
                  </DecodePageRoute>
                }
                path="/course/:courseId"
              />
              <Route
                element={
                  <DecodePageRoute>
                    <Courseware Container />
                  </DecodePageRoute>
                }
                path="/preview/course/:courseId/:sequenceId/:unitId"
              />
              <Route
                element={
                  <DecodePageRoute>
                    <Courseware Container />
                  </DecodePageRoute>
                }
                path="/preview/course/:courseId/:sequenceId"
              />
            </Routes>
          </div>
        </UserMessagesProvider>
      </NoticesProvider>
    </PathFixesProvider>
  </AppProvider>
</React Strict Mode>
`;
